src/network.rs

```diff
@@ -84,130 +91,79 @@ impl Network {
                         "section with prefix [{}] already exists",
                         prefix1
                     );
-
-                    let _ = self.sections.remove(&old_prefix);
                 }
-                Response::Reject(_) => {
-                    stats.rejections += 1;
-                }
-                Response::RelocateRequest {
-                    src,
-                    dst,
-                    node_name,
-                } => {
-                    let section = self.find_matching_section(dst);
-                    section.receive(Request::RelocateRequest {
-                        src,
-                        dst,
-                        node_name,
-                    })
-                }
-                Response::Relocate { dst, node } => {
-                    stats.relocations += 1;
-                    let section = self.find_matching_section(dst);
-                    section.receive(Request::Relocate(node))
-                }
-                Response::Send(prefix, request) => {
-                    match request {
-                        Request::Merge(target_prefix) => {
-                            // The receiver of `Merge` might not exists, because
-                            // it might have already split. So send the request
-                            // to every section with matching prefix.
-                            for section in self.sections.values_mut().filter(|section| {
-                                prefix.is_ancestor(&section.prefix())
-                            })
-                            {
-                                section.receive(Request::Merge(target_prefix))
-                            }
+                Action::Send(message) => {
+                    if let Message::Merge { source, target } = message {
+                        // Send the message to every descendant of `source` because
+                        // it might have already split.
+                        let mut found = false;
+                        for section in self.sections.values_mut().filter(|section| {
+                            section.prefix().is_descendant(&source)
+                        })
+                        {
+                            section.receive(Message::Merge { source, target });
+                            found = true;
                         }
-                        Request::Relocate(node) => {
-                            stats.relocations += 1;
-                            self.send(prefix, Request::Relocate(node));
+
+                        if !found {
+                            panic!("No descendant sections of {:?} found", source)
+                        }
+                    } else {
+                        let target = message.target();
+                        if let Some(section) = self.sections.values_mut().find(|section| {
+                            section.prefix().matches(target)
```

Comment:
for RelocateAccept and RelocateReject messages, the receiver is defined by node_name, not the target